File: week1/cpp/opening.cpp
/*
Copyright 2017 BIG VISION LLC ALL RIGHTS RESERVED

This code is made available to the students of 
the online course titled "Computer Vision for Faces" 
by Satya Mallick for personal non-commercial use. 

Sharing this code is strictly prohibited without written
permission from Big Vision LLC. 

For licensing and other inquiries, please email 
spmallick@bigvisionllc.com 
*/

#include <iostream>
#include <opencv2/highgui.hpp>
#include <opencv2/imgproc.hpp>

using namespace cv;
using namespace std;

int main() {
    Mat image;
    // Image read in and stored in Mat image
    image = imread("../data/images/opening.png", IMREAD_GRAYSCALE);
	
    // Initializing the number of iterations required
    int iterations = 3;
    // Create a structuring element
    int erosionSize = 3;
    // Selecting a elliptical kernel 
    Mat element = getStructuringElement(cv::MORPH_ELLIPSE,cv::Size(2 * erosionSize + 1, 2 * erosionSize + 1),cv::Point(erosionSize, erosionSize));
    // Mat imageMorphOpened stores the image after opening operation
    Mat imageMorphOpened;

    // Applying erosion followed by dilation on image by using morphologyEx function and specifying MORPH_OPEN tag(denotes opening operation) 
    morphologyEx(image, imageMorphOpened, MORPH_OPEN, element, Point(-1,-1),iterations);         
    
    // Display original image
    namedWindow("Original Image", CV_WINDOW_AUTOSIZE);
    imshow("Original Image", image);

    // Display the image after morphological opening is done
    namedWindow("Opening", CV_WINDOW_AUTOSIZE);
    imshow("Opening", imageMorphOpened);   

    // Wait for the user to press any key
    waitKey(0);
    return 0;
}

File: week1/python/opening.py
# Copyright 2017 BIG VISION LLC ALL RIGHTS RESERVED
# 
# This code is made available to the students of 
# the online course titled "Computer Vision for Faces" 
# by Satya Mallick for personal non-commercial use. 
#
# Sharing this code is strictly prohibited without written
# permission from Big Vision LLC. 
#
# For licensing and other inquiries, please email 
# spmallick@bigvisionllc.com 
#

import cv2

imageName = "../data/images/opening.png"
# Read the input image
image = cv2.imread(imageName, cv2.IMREAD_GRAYSCALE)

# Check for an invalid input
if image is None:  
    print("Could not open or find the image")

# Get structuring element/kernel which will be used for opening operation
openingSize = 3

# Select an elliptical kernel
element = cv2.getStructuringElement(cv2.MORPH_ELLIPSE, (2 * openingSize + 1, 2 * openingSize + 1),(openingSize,openingSize))

# Apply erosion on the image followed by dilation using morphologyEx function to specify MORPH_OPEN(denotes opening operation)
imageMorphOpened = cv2.morphologyEx(image, cv2.MORPH_OPEN, element,iterations=3)

# Display the original image
cv2.namedWindow("Original Image", cv2.WINDOW_AUTOSIZE)
cv2.imshow("Original Image", image)		

# Image after morphological opening operation
cv2.namedWindow("Opening", cv2.WINDOW_AUTOSIZE)
cv2.imshow("Opening",imageMorphOpened)		

# Wait for user to press any key
cv2.waitKey(0)
cv2.destroyAllWindows()
